pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,332 @@
+					From: corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: imageName,
+					},
+					To: &corev1.LocalObjectReference{Name: imageTag},
+					ReferencePolicy: imagev1.TagReferencePolicy{
+						Type: imagev1.LocalTagReferencePolicy,
+					},
+				}},
+			},
+		}
+
+		r.log.Debugf("creating imageStreamImport %v on cluster %s", imageStreamImport, clusterName)
+		// TODO (hongkliu): Add metrics for recording the results
+		if err := createAndCheckStatus(r.ctx, client, imageStreamImport); err != nil {
+			return err
+		}
+		log.Debug("Imported successfully")
+	}
+	return nil
+}
+
+func createAndCheckStatus(ctx context.Context, client ctrlruntimeclient.Client, imageStreamImport *imagev1.ImageStreamImport) error {
+	// ImageStreamImport is not an ordinary api but a virtual one that does the import synchronously
+	if err := client.Create(ctx, imageStreamImport); err != nil {
+		return fmt.Errorf("failed to import Image: %w", err)
+	}
+
+	// This should never be needed, but we shouldn't panic if the server screws up
+	if imageStreamImport.Status.Images == nil {
+		imageStreamImport.Status.Images = []imagev1.ImageImportStatus{{}}
+	}
+	if imageStreamImport.Status.Images[0].Image == nil {
+		return fmt.Errorf("imageStreamImport did not succeed: reason: %s, message: %s", imageStreamImport.Status.Images[0].Status.Reason, imageStreamImport.Status.Images[0].Status.Message)
+	}
+	return nil
+}
+
+func findNewest(isTags map[string]*imagev1.ImageStreamTag) string {
+	result := ""
+	var t metav1.Time
+	for clusterName, isTag := range isTags {
+		if t.Before(&isTag.Image.CreationTimestamp) {
+			t = isTag.Image.CreationTimestamp
+			result = clusterName
+		}
+	}
+	return result
+}
+
+const pullSecretName = "registry-cluster-pull-secret"
+
+func (r *reconciler) ensureImagePullSecret(namespace string, client ctrlruntimeclient.Client, log *logrus.Entry) error {
+	secret, mutateFn := r.pullSecret(namespace)
+	return upsertObject(r.ctx, client, secret, mutateFn, log)
+}
+
+// https://issues.redhat.com/browse/DPTP-1656?focusedCommentId=15345756&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15345756
+// E.g., ci-operator uses the release controller configuration to determine
+// the version of OpenShift we create from the ImageStream, so we need
+// to copy the annotation if it exists
+const releaseConfigAnnotationPrefix = "release.openshift.io"
+
+func imagestream(imageStream *imagev1.ImageStream) (*imagev1.ImageStream, crcontrollerutil.MutateFn) {
+	stream := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+	}
+	return stream, func() error {
+		for k, v := range imageStream.Annotations {
+			if stream.Annotations == nil {
+				stream.Annotations = map[string]string{}
+			}
+			if strings.HasPrefix(k, releaseConfigAnnotationPrefix) {
+				stream.Annotations[k] = v
+			}
+		}
+		stream.Spec.LookupPolicy.Local = true
```

Comment:
we should copy this field from whatever it is in the source imagestream

Reply:
Done.

Reply:
Done.